### PR TITLE
Mac osx system notification

### PR DESCRIPTION
**Pull request by josegomez**
This Implements the System Notifications for MacOSX.

It does this by implementing an AppleAppDelegate and subscribing to several notification events which apple sends out when 
Screen Locks
Screen Saver Starts
System Logs Off
System Reboots

Also implemented a polling Task which checks the HIDIdleTime (A system Maintained environment value which tracks how long since last mouse / keyboard input) and raises an event if it goes above the pre-set threshold.

### CODE REVIEW

File: SQRLDotNetClientUI/Platform/OSX/AppDelegate .cs [hunk 1]
```diff
@@ -21,9 +23,31 @@ namespace SQRLDotNetClientUI.Platform.OSX
     [Register("AppDelegate")]
     public class AppDelegate : NSApplicationDelegate
     {
+        public bool FinishedLaunching = false;
```
Comment (alexhauser) [inline SQRLDotNetClientUI/Platform/OSX/AppDelegate .cs]: This provokes the following warning: 

`'AppDelegate.FinishedLaunching' hides inherited member 'NSApplicationDelegate.FinishedLaunching(NSObject)'. Use the new keyword if hiding was intended.`

Can we get rid of this?

File: SQRLDotNetClientUI/Platform/OSX/AppDelegate .cs [hunk 2]
```diff
@@ -21,9 +23,31 @@ namespace SQRLDotNetClientUI.Platform.OSX
     [Register("AppDelegate")]
     public class AppDelegate : NSApplicationDelegate
```
Comment (alexhauser) [inline SQRLDotNetClientUI/Platform/OSX/AppDelegate .cs]: The file name has an extra space at the end: `AppDelegate .cs `

File: SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SQRLDotNetClientUI.Platform.OSX
+{
+    /// <summary>
+    /// Provides access to Windows system events that are relevant to the
```
Comment (alexhauser) [inline SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs]: Should read `... MacOS system events ...`

File: SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        private Task _pollTask = null;
+        private bool _screensaverDetected = false;
+        private bool _idleDetected = false;
+        private int _maxIdleSeconds = 60 * 15; // Set a sensible default, will be overwritten anyway
+
+        public event EventHandler<SystemEventArgs> Idle;
+        public event EventHandler<SystemEventArgs> Screensaver;
+        public event EventHandler<SystemEventArgs> Standby;
+        public event EventHandler<SystemEventArgs> SessionLogoff;
+        public event EventHandler<SystemEventArgs> SessionLock;
```
Comment (alexhauser) [inline SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs:46]: This gives `The event 'SystemEventNotifier.SessionLock' is never used`.
Can we either implement it or hide the warning using `#pragma warning disable 67` /  `#pragma warning restore 67`?

File: SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    public class SystemEventNotifier : ISystemEventNotifier
+    {
+
+        /// <summary>
+        /// Specifies the checking interval for the polling thread.
+        /// </summary>
+        private readonly int POLL_INTERVAL = 5000;
+
+
+        private CancellationTokenSource _cts = null;
```
Comment (alexhauser) [inline SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs:35]: You're not instantiating the `CancellationTokenSource` and you are not assigning the `CancellationToken`.
This should be part of your constructor:

```C#
_cts = new CancellationTokenSource();
_ct = _cts.Token;
```

Additionally, I would also add the destructor that cancels the polling task before garbage-collecting the object:

```C#
~SystemEventNotifier()
{
    // Cancel the polling task
    _cts.Cancel();
}
```

File: SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        private CancellationToken _ct;
+        private Task _pollTask = null;
+        private bool _screensaverDetected = false;
```
Comment (alexhauser) [inline SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs]: This gives `The field 'SystemEventNotifier._screensaverDetected' is assigned but its value is never used` and can probably be deleted?!

**Review by alexhauser — APPROVED** [review]
Great work, Jose! 👍

File: SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs (hunk 4 above)
Comment (josegomez) [reply SQRLDotNetClientUI/Platform/OSX/SystemEventNotifier.cs:46]: This was a typo (fixed)

**Comment by josegomez** [comment]
Fixed all the issues, will merge once testing is done.

**Pull request merged**